File: src/github.com/oniony/TMSU/storage/database/upgrade.go
// Copyright 2011-2017 Paul Ruane.

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package database

import (
	"database/sql"
	"github.com/oniony/TMSU/common"
	"github.com/oniony/TMSU/common/log"
)

// unexported

func upgrade(tx *sql.Tx) error {
	version := currentSchemaVersion(tx)

	log.Infof(2, "database schema has version %v, latest schema version is %v", version, latestSchemaVersion)

	if version == latestSchemaVersion {
		log.Infof(2, "schema is up to date")
		return nil
	}

	noVersion := schemaVersion{}
	if version == noVersion {
		log.Infof(2, "creating schema")

		if err := createSchema(tx); err != nil {
			return err
		}

		// still need to run upgrade as per 0.5.0 database did not store a version
	}

	log.Infof(2, "upgrading database")

	if version.LessThan(schemaVersion{common.Version{0, 5, 0}, 0}) {
		log.Infof(2, "renaming fingerprint algorithm setting")

		if err := renameFingerprintAlgorithmSetting(tx); err != nil {
			return err
		}
	}
	if version.LessThan(schemaVersion{common.Version{0, 6, 0}, 0}) {
		log.Infof(2, "recreating implication table")

		if err := recreateImplicationTable(tx); err != nil {
			return err
		}
	}
	if version.LessThan(schemaVersion{common.Version{0, 7, 0}, 0}) {
		log.Infof(2, "updating fingerprint algorithms")

		if err := updateFingerprintAlgorithms(tx); err != nil {
			return err
		}
	}
	if version.LessThan(schemaVersion{common.Version{0, 7, 0}, 1}) {
		log.Infof(2, "recreating version table")

		if err := recreateVersionTable(tx); err != nil {
			return err
		}
	}

	log.Infof(2, "updating schema version")
	if err := updateSchemaVersion(tx, latestSchemaVersion); err != nil {
		return err
	}

	return nil
}

func renameFingerprintAlgorithmSetting(tx *sql.Tx) error {
	if _, err := tx.Exec(`
UPDATE setting
SET name = 'fileFingerprintAlgorithm'
WHERE name = 'fingerprintAlgorithm'`); err != nil {
		return err
	}

	return nil
}

func recreateImplicationTable(tx *sql.Tx) error {
	if _, err := tx.Exec(`
ALTER TABLE implication
RENAME TO implication_old`); err != nil {
		return err
	}

	if err := createImplicationTable(tx); err != nil {
		return err
	}

	if _, err := tx.Exec(`
INSERT INTO implication
SELECT tag_id, 0, implied_tag_id, 0
FROM implication_old`); err != nil {
		return err
	}

	if _, err := tx.Exec(`
DROP TABLE implication_old`); err != nil {
		return err
	}

	return nil
}

func updateFingerprintAlgorithms(tx *sql.Tx) error {
	rows, err := tx.Query(`
SELECT value
FROM setting
WHERE name = 'fileFingerprintAlgorithm'`)
	if err != nil {
		return err
	}
	defer rows.Close()

	var fileFingerprintAlg string
	if rows.Next() && rows.Err() == nil {
		rows.Scan(&fileFingerprintAlg) // ignore errors
	}

	switch fileFingerprintAlg {
	case "symlinkTargetName":
		if _, err := tx.Exec(`
INSERT INTO setting (name, value)
VALUES ("symlinkFingerprintAlgorithm", "targetName")`); err != nil {
			return err
		}
		if _, err := tx.Exec(`
DELETE FROM setting WHERE name = 'fileFingerprintAlgorithm';`); err != nil {
			return err
		}
	case "symlinkTargetNameNoExt":
		if _, err := tx.Exec(`
INSERT INTO setting (name, value)
VALUES ("symlinkFingerprintAlgorithm", "targetNameNoExt")`); err != nil {
			return err
		}
		if _, err := tx.Exec(`
DELETE FROM setting WHERE name = 'fileFingerprintAlgorithm';`); err != nil {
			return err
		}
	}

	return nil
}

func recreateVersionTable(tx *sql.Tx) error {
	if _, err := tx.Exec("DROP TABLE version;"); err != nil {
		return err
	}

	if err := createVersionTable(tx); err != nil {
		return err
	}

	if err := insertSchemaVersion(tx, latestSchemaVersion); err != nil {
		return err
	}

	return nil
}

File: src/storage/upgrade.rs
use crate::errors::*;
use crate::storage::schema::{self, SchemaVersion, LATEST_SCHEMA_VERSION};
use crate::storage::Transaction;

pub fn upgrade(tx: &mut Transaction) -> Result<()> {
    let current_version = schema::current_schema_version(tx)?;

    info!(
        "Database schema has version {:?}, latest schema version is {}",
        current_version, LATEST_SCHEMA_VERSION
    );

    if let Some(LATEST_SCHEMA_VERSION) = current_version {
        info!("Schema is up to date");
        return Ok(());
    }

    if current_version.is_none() {
        info!("Creating schema");
        schema::create_schema(tx)?;

        // Still need to run the upgrade, as per 0.5.0 the database did not store a version
    }

    // Use a fake 0.0.0-0 version as fallback for comparisons
    let current_version = current_version.unwrap_or(SchemaVersion::from_tuple(0, 0, 0, 0));

    info!("Upgrading database");

    if current_version < SchemaVersion::from_tuple(0, 5, 0, 0) {
        info!("Renaming fingerprint algorithm setting");
        rename_fingerprint_algorithm_setting(tx)?;
    }

    if current_version < SchemaVersion::from_tuple(0, 6, 0, 0) {
        info!("Recreating implication table");
        recreate_implication_table(tx)?;
    }

    if current_version < SchemaVersion::from_tuple(0, 7, 0, 0) {
        info!("Updating fingerprint algorithms");
        update_fingerprint_algorithms(tx)?;
    }

    if current_version < SchemaVersion::from_tuple(0, 7, 0, 1) {
        info!("Recreating version table");
        recreate_version_table(tx)?;
    }

    info!("Updating schema version");
    schema::update_schema_version(tx, &LATEST_SCHEMA_VERSION)?;

    Ok(())
}

fn rename_fingerprint_algorithm_setting(tx: &mut Transaction) -> Result<()> {
    tx.execute(
        "
UPDATE setting
SET name = 'fileFingerprintAlgorithm'
WHERE name = 'fingerprintAlgorithm'",
    )?;

    Ok(())
}

fn recreate_implication_table(tx: &mut Transaction) -> Result<()> {
    tx.execute(
        "
ALTER TABLE implication
RENAME TO implication_old",
    )?;

    schema::create_implication_table(tx)?;

    tx.execute(
        "
INSERT INTO implication
SELECT tag_id, 0, implied_tag_id, 0
FROM implication_old",
    )?;

    tx.execute("DROP TABLE implication_old")?;

    Ok(())
}

fn update_fingerprint_algorithms(tx: &mut Transaction) -> Result<()> {
    tx.execute(
        "
UPDATE setting
SET name = 'symlinkFingerprintAlgorithm',
    value = 'targetName'
WHERE name = 'fileFingerprintAlgorithm'
AND value = 'symlinkTargetName'",
    )?;

    tx.execute(
        "
UPDATE setting
SET name = 'symlinkFingerprintAlgorithm',
    value = 'targetNameNoExt'
WHERE name = 'fileFingerprintAlgorithm'
AND value = 'symlinkTargetNameNoExt'",
    )?;

    Ok(())
}

fn recreate_version_table(tx: &mut Transaction) -> Result<()> {
    tx.execute("DROP TABLE version")?;

    schema::create_version_table(tx)?;
    schema::insert_schema_version(tx, &LATEST_SCHEMA_VERSION)?;

    Ok(())
}
